tools/analyze_results.py

```diff
@@ -0,0 +1,176 @@
+
+
+class ResultEvaluate(object):
+
+    def __init__(self,
+                 dataset,
+                 results,
+                 topk=2,
+                 show=False,
+                 show_dir='work_dir'):
+        self.topk = topk
+        self.dataset = dataset
+        assert 0 < 2 * self.topk < len(self.dataset)
+        self.results = results
+        self.img_prefix = dataset.img_prefix
+        self.CLASSES = self.dataset.CLASSES
+        self.show = show
+        self.show_dir = show_dir
+        self.data_infos = [
+            self.dataset.data_infos[i] for i in range(len(self.dataset))
+        ]
+        self.annotations = [
+            self.dataset.get_ann_info(i) for i in range(len(self.dataset))
+        ]
+
+    def _show_image_gts_results(self, index, value, show=False, out_dir=None):
+        gts = self.annotations[index]
+        data_info = self.data_infos[index]
+        result = self.results[index]
+        if self.img_prefix is not None:
+            filename = osp.join(self.img_prefix, data_info['filename'])
+        else:
+            filename = data_info['filename']
+        img = mmcv.imread(filename)
+        # TODO gt mask support
+        imshow_det_bboxes(
+            img,
+            gts['bboxes'],
+            gts['labels'],
+            class_names=self.CLASSES,
+            bbox_color=(255, 102, 61),
+            text_color=(255, 102, 61),
+            show=False)
+        if isinstance(result, tuple):
+            bbox_result, segm_result = result
+            if isinstance(segm_result, tuple):
+                segm_result = segm_result[0]
+        else:
+            bbox_result, segm_result = result, None
+        bboxes = np.vstack(bbox_result)
+        labels = [
+            np.full(bbox.shape[0], i, dtype=np.int32)
+            for i, bbox in enumerate(bbox_result)
+        ]
+        labels = np.concatenate(labels)
+
+        fname, name = osp.splitext(osp.basename(filename))
+        save_filename = fname + '_' + str(round(value, 3)) + name
+        out_file = osp.join(out_dir, save_filename)
+        imshow_det_bboxes(
+            img,
+            bboxes,
+            labels,
+            segms=segm_result,
+            class_names=self.CLASSES,
+            bbox_color=(72, 101, 241),
+            text_color=(72, 101, 241),
+            show=show,
+            out_file=out_file)
+
+    def _save_images(self, mAPs, save_dir):
+        mmcv.mkdir_or_exist(save_dir)
+        for mAP in mAPs:
+            self._show_image_gts_results(mAP[0], mAP[1], self.show, save_dir)
+
+    def _eval_fn(self, det_result, annotation):
+        iou_thrs = np.linspace(
+            .5, 0.95, int(np.round((0.95 - .5) / .05)) + 1, endpoint=True)
+        mean_aps = []
+        for thr in iou_thrs:
+            mean_ap, _ = eval_map([det_result], [annotation], iou_thr=thr)
+            mean_aps.append(mean_ap)
+        return sum(mean_aps) / len(mean_aps)
+
+    def evaluate(self):
```

Comment:
Can we make the method a method of the dataset? Thus for each dataset, we can call it using `dataset.visualize`